### PR TITLE
Migrating SkyVerge API and Exponential Backoff (5/10)

### PR DESCRIPTION
**Pull request by message-dimke**
Ref #2095 Implement Exponential Backoff preventing issues caused by unavailable API services.

Series of 10 PRs to migrate away from **SkyVerge API** into internal **WC_Facebookcommerce_Graph_API** and get rid of SkyVerge dependency.

### Changelog entry

> Tweak - Moving from SkyVerge Facebook Graph API into the internal one.

### CODE REVIEW

File: facebook-commerce.php [hunk 1]
```diff
@@ -597,81 +597,73 @@ public function get_graph_api() {
...
 	 * @param string $product_group_id product group ID
-	 * @return array
+	 * @return array a map of ( `retailer id` -> `id` ) pairs.
 	 */
-	private function find_variation_product_item_ids( $product_group_id ) {
-
-		$product_item_ids = array();
-
+	private function find_variation_product_item_ids( string $product_group_id ): array {
+		$product_item_ids = [];
```
Comment (rawdreeg) [inline facebook-commerce.php:642]: For consistency's sake, we'd probably have to stick to one syntax style for the arrays. Can we change this to array to keep the consistency in this extension perhaps?

PS:  I see we use the short array syntax in some of the other extensions but on this extension, we still use array().  Maybe there's a bigger discussion to be had with regard to standardizing this.

**Review by rawdreeg — APPROVED** [review]
✅  This looks good. I have tested the changes and everything works well.

I left a comment regarding an array declaration that used the shorthand syntax. We might wanna keep this consistent throughout the extension. Other than that, I am approving the changes.

File: facebook-commerce.php (hunk 1 above)
Comment (message-dimke) [reply facebook-commerce.php:642]: Adjusted. Thanks for noticing.

**Review by rawdreeg — APPROVED** [review]

**Comment by message-dimke** [comment]
## PR (1/10)

### Changes proposed in this Pull Request:

Removing `get_page`, `get_page_name` and `get_page_url` API methods as the ones not used. Migrating SkyVerge endpoint to retrieve product group product ids. Implementing `next` api method to enable pagination for multipage responses.

### Detailed test instructions:

1. Create new or reuse any existing Variable Product
2. If Product is not set to sync with Facebook - set it to do so, or, of Product is already synced - add a new variation and set it to sync with Facebook.

Edit_product_“V-Neck_T-Shirt_Category_Comma_Compatible”_‹_WordPress-Facebook_—_WordPress

3. Take a note of Product variants synced with Facebook by the time of this test.

Edit_product_“V-Neck_T-Shirt_Category_Comma_Compatible”_‹_WordPress-Facebook_—_WordPress

or if none

Edit_product_“Hoodie”_‹_WordPress-Facebook_—_WordPress

4. Hit Update Product button to save changes and sync the product.
5. Check new Product variation got its corresponding id 

Edit_product_“V-Neck_T-Shirt_Category_Comma_Compatible”_‹_WordPress-Facebook_—_WordPress

or, if were none

Edit_product_“Hoodie”_‹_WordPress-Facebook_—_WordPress

**Comment by message-dimke** [comment]
## PR (2/10)

### Changes proposed in this Pull Request:

Migrating SkyVerge `get_catalog` endpoint to retrieve catalog name for Facebook. Adjusting response parsing and `get_data` shortcut methods.

### Detailed test instructions:

1. Connect/reconnect to Facebook or do noting, since Facebook Catalog name is fetched on the fly w/o any caching or storing in the database.
2. See Facebook Catalog name is displayed

Facebook_for_WooCommerce_‹_WordPress-Facebook_—_WordPress

**Comment by message-dimke** [comment]
## PR (3/10)

### Changes proposed in this Pull Request:

Migrating SkyVerge `get_user` and `delete_user_permission` endpoints to disconnect from Facebook.

### Detailed test instructions:

1. Disconnect from Facebook.

Facebook_for_WooCommerce_‹_WordPress-Facebook_—_WordPress

2. Observe your store disconnected successfully.

**Comment by rawdreeg** [comment]
@message-dimke, I tested the latest changes. 👍🏾 Looking good.!

**Review by rawdreeg — APPROVED** [review]

**Comment by message-dimke** [comment]
## PR (3/10)

### Changes proposed in this Pull Request:

Migrating SkyVerge `get_user` and `delete_user_permission` endpoints to disconnect from Facebook.

### Detailed test instructions:

1. Disconnect from Facebook.

Facebook_for_WooCommerce_‹_WordPress-Facebook_—_WordPress

2. Observe your store disconnected successfully.

**Review by rawdreeg — APPROVED** [review]
✅  LGTM!

**Comment by message-dimke** [comment]
## PR (4/10)

### Changes proposed in this Pull Request:

Migrating SkyVerge `send_item_updates` endpoints [Product Catalog Items Batch](https://developers.facebook.com/docs/marketing-api/reference/product-catalog/items_batch/).

### Detailed test instructions:

1. Go to Products > All Products and select product to update.

Products_‹_WordPress-Facebook_—_WordPress

2.  Pick Edit from bulk actions and hit Apply.
3. Update some property. e.g. `In Stock?` and hit Update button

Products_‹_WordPress-Facebook_—_WordPress

4. Go to WooCommerce Status page and open Logs tab

WooCommerce_status_‹_WordPress-Facebook_—_WordPress

WooCommerce_status_‹_WordPress-Facebook_—_WordPress

P.S. to see the logs, make sure to enable logging here

Facebook_for_WooCommerce_‹_WordPress-Facebook_—_WordPress

**Review by rawdreeg — APPROVED** [review]
I reviewed and test the last set of changes. Looks good. Thanks 👍🏾

**Comment by message-dimke** [comment]
## PR (5/10)

### Changes proposed in this Pull Request:

Migrating SkyVerge `send_pixel_events`.

### Detailed test instructions:

1. Go to any product page.

Long_Sleeve_Tee_–_WordPress-Facebook

2. Open Facebook Events Manager.

Events_Manager

3. View Page Views.

Events_Manager

4. Open View Details and see recent events.

Events_Manager

5. Replay the same steps for View Content recent events.

Events_Manager

File: includes/API.php [hunk 2]
```diff
@@ -481,26 +480,6 @@ public function delete_product_item( $product_item_id ) {
...
-	 * Sends Pixel events.
-	 *
-	 * @since 2.0.0
-	 *
-	 * @param string  $pixel_id pixel ID
-	 * @param Event[] $events events to send
-	 * @return Response
-	 * @throws Framework\SV_WC_API_Exception
-	 */
-	public function send_pixel_events( $pixel_id, array $events ) {
```
Comment (rawdreeg) [inline includes/API.php:598]: This is a public method. should we perhaps deprecate it before a complete removal?

**Review by rawdreeg — APPROVED** [review]
I have reviewed and tested the changes. It seems good to me, except for a question regarding removing or changing the public method namespace. should be fine but I thought It was worth some consideration

File: includes/API.php (hunk 2 above)
Comment (rawdreeg) [reply includes/API.php:598]: the likelihood of this being used outside of this extension is pretty low. so I reckon this is okay.

**Comment by message-dimke** [comment]
> I have reviewed and tested the changes. It seems good to me, except for a question regarding removing or changing the public method namespace. should be fine but I thought It was worth some consideration

What an interesting thought 👍 
I think we revert all the changes to `SkyVerge\WooCommerce\Facebook\API` since most of apis are public and I did remove all of them after migrating.

**Comment by rawdreeg** [comment]
> What an interesting thought 👍 I think we revert all the changes to `SkyVerge\WooCommerce\Facebook\API` since most of apis are public and I did remove all of them after migrating.

After giving this more thought, I reckon it's probably okay. It's very unlikely that these methods would be called outside of this extension.

**Comment by message-dimke** [comment]
Feature branch created.
Code moved into https://github.com/woocommerce/facebook-for-woocommerce/pull/2212

**Pull request closed without merging**